app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,202 @@
+        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
+      }),
+    )
+  }
+
+  const firstValidRPCUrl = Array.isArray(rpcUrls)
+    ? rpcUrls.find((rpcUrl) => validUrl.isHttpsUri(rpcUrl))
+    : null
+
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          validUrl.isHttpsUri(blockExplorerUrl),
+        )
+      : null
+
+  if (!firstValidRPCUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+      }),
+    )
+  }
+
+  if (blockExplorerUrls !== null && !firstValidBlockExplorerUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected null or array with at least one valid string HTTPS URL 'blockExplorerUrl'. Received: ${blockExplorerUrls}`,
+      }),
+    )
+  }
+
+  const _chainId = typeof chainId === 'string' && chainId.toLowerCase()
+
+  if (customRpcExistsWith({ rpcUrl: firstValidRPCUrl, chainId: _chainId })) {
+    return end(
+      ethErrors.rpc.internal({
+        message: `Ethereum chain with the given RPC URL and chain ID already exists.`,
+        data: { rpcUrl: firstValidRPCUrl, chainId },
+      }),
+    )
+  }
+
+  if (!isPrefixedFormattedHexString(_chainId)) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected 0x-prefixed, unpadded, non-zero hexadecimal string 'chainId'. Received:\n${chainId}`,
+      }),
+    )
+  }
+
+  if (!isSafeChainId(parseInt(_chainId, 16))) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Invalid chain ID "${_chainId}": numerical value greater than max safe value. Received:\n${chainId}`,
+      }),
+    )
+  }
+
+  let endpointChainId
+
+  try {
+    endpointChainId = await jsonRpcRequest(firstValidRPCUrl, 'eth_chainId')
+  } catch (err) {
+    return end(
+      ethErrors.rpc.internal({
+        message: `Request for method 'eth_chainId on ${firstValidRPCUrl} failed`,
+        data: { networkErr: err },
+      }),
+    )
+  }
+
+  if (_chainId !== endpointChainId) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Chain ID returned by RPC URL ${firstValidRPCUrl} does not match ${_chainId}`,
+        data: { chainId: endpointChainId },
+      }),
+    )
+  }
+
+  if (typeof chainName !== 'string' || !chainName) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected non-empty string 'chainName'. Received:\n${chainName}`,
+      }),
+    )
+  }
+  const _chainName =
+    chainName.length > 100 ? chainName.substring(0, 100) : chainName
+
+  if (nativeCurrency !== null && typeof nativeCurrency !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected null or object 'nativeCurrency'. Received:\n${nativeCurrency}`,
+      }),
+    )
+  }
```

Comment:
nit: the `typeof nativeCurrency !== 'object'` can be brought into the `if` block currently starting on line 156.

I also think the complete check should be: 

`typeof nativeCurrency !== 'object' || Array.isArray(nativeCurrency)`